auth/standard_role_manager.cc

```diff
@@ -250,26 +251,32 @@ future<> standard_role_manager::start() {
             }
             return make_ready_future<>();
         }).then([this] {
-            _stopped = auth::do_after_system_ready(_as, [this] {
-                return seastar::async([this] {
+            _stopped = auth::do_after_system_ready(_as, [this] () -> future<>  {
+                auto signal_super_user_created = defer([this] {
+                    _superuser_created.signal();
+                });
+                try {
                     if (legacy_mode(_qp)) {
-                        _migration_manager.wait_for_schema_agreement(_qp.db().real_database(), db::timeout_clock::time_point::max(), &_as).get();
+                        co_await _migration_manager.wait_for_schema_agreement(_qp.db().real_database(), db::timeout_clock::time_point::max(), &_as);
 
-                        if (any_nondefault_role_row_satisfies(_qp, &has_can_login).get()) {
+                        if (co_await any_nondefault_role_row_satisfies(_qp, &has_can_login)) {
                             if (legacy_metadata_exists()) {
                                 log.warn("Ignoring legacy user metadata since nondefault roles already exist.");
                             }
-
-                            return;
+                            co_return;
                         }
 
                         if (legacy_metadata_exists()) {
-                            migrate_legacy_metadata().get();
-                            return;
+                            co_await migrate_legacy_metadata();
+                            co_return;
                         }
                     }
-                    create_default_role_if_missing().get();
-                });
+                    co_await create_default_role_if_missing();
+                } catch (...) {
+                    // If we fail to create the default role, we should not signal that we are ready.
```

Comment:
cool, but then nobody will signal and code in main will block forever, since future is awaited only during stop() method nobody before that will see the error as well

please manually test by inserting throw into create_default_role_if_missing(), code is not trivial and it's easy to write something which wouldn't handle errors properly

Reply:
you need to attach try catch block to future returned by do_after_system_ready because it throws when it's done with retries, pseudocode:

```
auth::do_after_system_ready(...).then_wrapped(future<> f) {
try {
f.get()
catch(...) { ...broken(...); throw; }
...signal()
});
```